mux/mux.go

```diff
@@ -3,104 +3,87 @@ package mux
 import (
 	"context"
 	"errors"
-	"net"
+	"log"
 	"net/http"
+	"strings"
 	"time"
 
-	"github.com/soheilhy/cmux"
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
 	"google.golang.org/grpc"
 )
 
 const defaultGracePeriod = 5 * time.Second
+const grpcRequestMarker = "application/grpc"
 
-// Serve starts a TCP listener and serves the registered protocol
-// servers (i.e., http, grpc, etc.) and blocks until server exits.
-// Context can be cancelled to perform graceful shutdown.
-// Use WithHTTP() and WithGRPC() to register procol servers.
+// Serve starts a TCP listener and serves the registered protocol servers and blocks
+// until server exits. Context can be cancelled to perform graceful shutdown.
+// Use WithHTTP() and WithGRPC() to register protocol servers.
 func Serve(ctx context.Context, listenAddr string, opts ...Option) error {
-	var mux cmuxWrapper
+	var mux protocolMuxServer
 	for _, opt := range opts {
 		if err := opt(&mux); err != nil {
 			return err
 		}
 	}
-	if mux.httpServer == nil && mux.grpcServer == nil {
+	if mux.httpHandler == nil && mux.grpcServer == nil {
 		return errors.New("at-least one of http & grpc server must be set")
 	}
-
 	return mux.Serve(ctx, listenAddr)
 }
 
-type cmuxWrapper struct {
-	httpServer  *http.Server
+type protocolMuxServer struct {
+	httpHandler http.Handler
 	grpcServer  *grpc.Server
 	gracePeriod time.Duration
 }
 
-func (cmw *cmuxWrapper) Serve(baseCtx context.Context, addr string) error {
+func (pmux *protocolMuxServer) Serve(baseCtx context.Context, addr string) error {
 	ctx, cancel := context.WithCancel(baseCtx)
 	defer cancel()
 
-	l, err := net.Listen("tcp", addr)
-	if err != nil {
-		return err
+	http1Server := &http.Server{
```

Comment:
why is `h2c` handler named as `http1Server`?

Reply:
It's not. this is the standard `http.Server` type

Reply:
but aren't we using `h2c` and not `http` ?
```go
httpServer := &http.Server{
		Addr:    addr,
		Handler: h2c.NewHandler(pmux.muxedHandler(), &http2.Server{}),
}
```

also, why `h2c` ?

Reply:
h2c is hosted on http1. It's simply a handler. If http2 negotiation works out, the underlying connection is hijacked and upgraded for h2c .